docs/network/wsProtocol.md

```diff
@@ -113,6 +113,21 @@ A successful push operation will usually result in two messages from the server
 
 If the server staggers the update messages, it is fine to send the update message including more changes that the latest push, but it must still be sent before the pushResult message. If necessary, the server may artificially delay the `pushResult` message on the server for up to 5-10 seconds.
 
+## Authentication
+
+Oncilla supports a first class `auth` action, designed to establish an initial/ongoing authenticated websocket session with the server. The message is structured as follows:
+
+```json
+{
+  "action": "auth",
+  "token": ""
+}
+```
+
+The purpose of this action is to establish/re-establish authentication with the Server
+
```

Comment:
>The server may respond with an `authResult` message

Would it be better if the server _must_ respond to the auth message? If an app doesn't need a response the server could return an empty result and the client could throw it away. (My guess is any robust app would at least want a response to be able to stop a retry/failure timer.) From a protocol perspective providing a symmetrical `<action>Result` for any given `action` would be a very consistent and predictable request/response pattern. (It would make sense to deviate from this for pub/sub, of course.)

Reply:
I would like us to avoid that requirement.

For one, a few lines above Oncilla claims it’s indifferent to the authentication process, and here it would suddenly change.

More importantly for me, though, this would contradict the spirit of the design, which is focused on use cases and not on technical implementation. You could argue for the tradeoff the other way, but my preference is for the former.

Finally, I personally don’t see a lot of value in the request/response pattern. This is not HTTP, we’re sending minimal data on a per-need basis. 🤷‍♂

Reply:
>I personally don’t see a lot of value in the request/response pattern. This is not HTTP

This may not be HTTP, but there are plenty of protocols that utilize persistent connections that implement a request/response pattern. [ZeroMQ](http://zeromq.org/) is one good example that I feel could offer guidance in this area. I totally support the library focusing on use cases over technical implementation as well. I'm just trying to say that the more consistent a library is with its interfaces and behavior, the easier and less error-prone it is for a third-party developer to integrate with that library. That being said, there are certainly times for deviation.

>I would like us to avoid that requirement.
>[...] a few lines above Oncilla claims it’s indifferent to the authentication process

Fair enough, no problem. I would be happy to approve the optional response as well.

Reply:
Just another thought here, in the case we didn't want to go down the route of an `authResponse`, my expectation would be that the server send a termination message to the client and disconnect the socket on auth failure. That's another explicit way to inform the client if the auth succeeded/failed without needing to [optionally] introduce another message type.

Reply:
(Alex asked me to chime in to this convo since he knows I have a love of protocol design - my apologies if I'm an unwanted cook in the kitchen, but I'm just chiming in with my thoughts as requested; also, please correct me if I'm misunderstanding some part of the situation due to not having full context)

Making this response optional seems wonky to me, and I'll present my reasoning on that as a sort of decision tree:
- If *no* client implementations will have a hard dependency on this response, then it should be omitted entirely from the protocol. There is no need for having a response just for the sake of aesthetic symmetricality - as @denis-sokolov rightly pointed out.
- If *some or all* client implementations will have a hard dependency on this response, then they should have a way to guarantee that any protocol-compliant server will meet their need. This implies one of the following options:
  - The server should be required to send the response (optimizing for simplicity of the protocol)
  - The client should have some way of explicitly asking the server to send the response, and the server should then be required to send it. This complicates the protocol, but does optimize for reducing unwanted traffic when the response is not needed. 
    - For example, there could be a parameter as part of the `auth` message from the client called `responseNeeded`, or something to that effect, which would tell the server that it must send the response.
    - Alternatively, there could be a request message like `getAuthState`, or something to that effect, which a client could use to know whether their connection is currently authenticated, for clients that care. Again, if no clients actually care about this, it shouldn't be in the protocol to begin with.

Reply:
In the event that you do decide for some reason to keep it as optional on the server side with no control from the client side, I recommend at least making the protocol more explicit that it is the prerogative of the server to decide whether to send the response. Otherwise, I would read the language suggested in Denis' comment ("The server may respond") and wonder what "may" means in this context - under what circumstances is such a response sent? Language should be added to indicate that the circumstances are arbitrary and unknown from the client's perspective - that it is strictly the perogative of the server.

Again, when I elaborate this language explicitly it reinforces to me that it seems wonky for a protocol to work this way.

Reply:
Thanks for the feedback @jemc.

>if no clients actually care about this, it shouldn't be in the protocol to begin with.

I believe the project we're currently integrating with this library does require the client to know the authentication state, since all users must be authenticated to access the backend's data.

Reply:
I think I need to clarify my understanding of what indifference to auth means for oncilla protocol. Designing a particular authentication creates a new protocol which extends ours. This makes both the client and server not universal anymore: they can’t talk to other protocol-compliant clients and servers.

For a thorough example, let’s imagine I am building an application A, where the client does not need the authResult message. Let’s imagine John is building application B, where the client needs an authResult message with a value of "success" | "wrong-password" | "wrong-email".

If we add this requirement that the server needs to send an authResult message, my server A will need to send the message back to be valid. But since I’m building client A, which does not care about the message, my authResult message will be empty, just a dance to comply with the protocol.

Now imagine client B trying to talk to server A. Client B receives an empty authResult message from server A and crashes, because the contents of the message are not valid in the world of client B.

So what was the benefit of making server A send that message?

The project we’re currently integrating with does not require the authResult message, it’s application A in the example above. In the regular flow the token the client sends is already validated by Auth0 and the client knows for sure it will work. The client will not ever receive any meaningful output in the authResult message. If something unexpected happens and the server decides to reject a token, it will close the connection, like it would if the token expires in the middle of the session.

If you guys want to further continue this, let’s jump on a call, because this is getting too long of a comment thread. :-)